File: Demos/OC/UFileSDKDemo/UFileSDKDemo/file/UFDataManager.m
//
//  UFDataManager.m
//  UFileAssistant
//
//  Created by ethan on 2018/11/15.
//  Copyright © 2018 ucloud. All rights reserved.
//

#import "UFDataManager.h"
@interface UFDataManager()
@property (nonatomic,strong) NSData *fileData;
@property (nonatomic,assign) NSInteger fileSize;
@property (nonatomic,assign) NSInteger allParts;
@end

@implementation UFDataManager

- (instancetype)initUFDataManagerWithUFMultiPartInfo:(UFMultiPartInfo *)multiPartInfo filePath:(NSString *)filePath 
{
    self = [super init];
    if (!self) {
        return nil;
    }
    
    self.multiPartInfo = multiPartInfo;
    self.filePath = filePath;
    self.fileData = [NSData dataWithContentsOfFile:self.filePath];
    NSDictionary *fileAttributes =  [[NSFileManager defaultManager] attributesOfItemAtPath:self.filePath error:nil];
    self.fileSize = (NSInteger)[fileAttributes fileSize];
    _allParts = (self.fileSize + self.multiPartInfo.blkSize - 1)/self.multiPartInfo.blkSize;

    self.etags = [NSMutableDictionary dictionary];
    return self;
}

- (NSInteger)allParts
{
    return _allParts;
}

- (void)resetTableData
{
    _multiPartInfo = nil;
    _filePath = NULL;
    _etags = [NSMutableDictionary dictionary];
    _allParts = 0;
}

- (NSData *)getDataForPart:(NSInteger)partNumber
{
    if (partNumber >= self.allParts) {
        return nil;
    }
    
    NSInteger loc = partNumber*self.multiPartInfo.blkSize;
    NSInteger  length = self.multiPartInfo.blkSize;
    NSInteger totalen = loc + length;
    if (totalen > self.fileSize) {
        length = self.fileSize - loc;
    }
    return [self.fileData subdataWithRange:NSMakeRange(loc, length)];
}

- (NSString *)writeData:(NSData *)data fileName:(NSString *)fileName {
    NSString *tempDirectory = NSTemporaryDirectory();
    NSString *tempDataDirectory = [tempDirectory stringByAppendingPathComponent:@"tempData"];
    // 创建文件夹
    NSError *error;
    NSString *filePath;
    BOOL success = [[NSFileManager defaultManager] createDirectoryAtPath:tempDataDirectory
                                             withIntermediateDirectories:YES
                                                              attributes:nil
                                                                   error:&error];
    if (success) {
        // 创建文件路径
        filePath = [tempDataDirectory stringByAppendingPathComponent:fileName];
        // 将数据写入文件
        success = [data writeToFile:filePath options:NSDataWritingAtomic error:&error];
        if (success) {
            return filePath;
        }
    }
    return nil;
}

- (void)addEtag:(NSString *)etag partNumber:(NSInteger)partNumber
{
    [self.etags setObject:etag forKey:[NSString stringWithFormat:@"%ld",partNumber]];
}
@end

File: Demos/Swift/UFileSDKDemo-swift/UFileSDKDemo-swift/file/UFDataManager.swift
//
//  UFDataManager.swift
//  UFileSDKDemo-swift
//
//  Created by ethan on 2018/12/12.
//  Copyright © 2018 ucloud. All rights reserved.
//

import UIKit

class UFDataManager {

    var multiPartInfo:UFMultiPartInfo?
    var filePath:String?
    var etags:NSMutableDictionary
    
    var fileData:NSData? = nil
    var fileSize:UInt = 0
    var allParts:UInt = 0
    
    init(multiPartInfo:UFMultiPartInfo, filePath:String) {
        
        self.multiPartInfo = multiPartInfo
        self.filePath = filePath
        self.etags = NSMutableDictionary()
        do{
        self.fileData = try NSData(contentsOfFile: self.filePath!)
        let attr = try FileManager.default.attributesOfItem(atPath: self.filePath!)
        let dict = attr as NSDictionary
            self.fileSize = UInt(dict.fileSize())
            self.allParts = (self.fileSize + self.multiPartInfo!.blkSize - 1)/self.multiPartInfo!.blkSize
            
        }catch{
            print("Error:\(error)")
        }
    }
    
    func resetTableData() -> Void
    {
        multiPartInfo = nil
        filePath = nil
        etags = NSMutableDictionary()
        allParts = 0
    }
    
    func getDataForPart(partNumber:UInt) -> Data
    {
        if partNumber >= allParts {
            return Data()
        }
        
        let loc  = partNumber*self.multiPartInfo!.blkSize
        var length  = self.multiPartInfo!.blkSize
        let totalen = loc + length
        if totalen > self.fileSize {
            length = self.fileSize - loc
        }
        return NSMutableData.init(data: self.fileData!.subdata(with: NSRange.init(location:Int(loc), length: Int(length)))) as Data
    }
    
    func addEtag(etag:String, partNumber:Int)
    {
        self.etags .setValue(etag, forKey: String(partNumber))
    }
}
